### PR TITLE
Clarify knn_graph data ordering for user-provided input

### PR DESCRIPTION
**Pull request by elisno**
When Datalab constructs the `knn_graph: csr_matrix` from features, it adheres to a specific data ordering which assumed to hold in existing algorithms specialised on knn-search. This ordering deviates from the standard canonical format of arbitrary CSR-formatted matrices where indices are sorted by column within each row. Similarly, the `kneighbors_graph` function in scikit-learn also follows this unique ordering. However, when users provide their own `knn_graph: csr_matrix`, they may not be aware of this requirement. This commit adds clarification in the documentation to ensure users follow the correct data ordering—sorting values, not columns, within each row—when providing a custom `knn_graph: csr_matrix`.

## Summary

🎯 **Purpose**: 

The purpose of this PR is to clarify the documentation regarding the data ordering requirements for user-provided `knn_graph: csr_matrix` input to Datalab. This change aims to prevent potential user errors that could arise from a misunderstanding of the specific data ordering Datalab adheres to when constructing the `knn_graph: csr_matrix` from features, especially as this ordering doesn't follow the canonical format of CSR matrices that is more general.

## Impact

 🌐 Areas Affected: Documentation of knn_graph: csr_matrix input.
 👥 Who’s Affected: Users providing their own `knn_graph: csr_matrix` input.

## Testing

Documentation clarification; no code change to test.

## References

The data ordering adhered to by Datalab when constructing the knn_graph: csr_matrix from features aligns with the ordering followed by the `kneighbors_graph` function in scikit-learn ([link](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.kneighbors_graph.html)), which is crucial for the correct functioning of algorithms specialized on knn-search.

Here are the docs for `csr_matrix` in scipy ([link](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html)).

## Reviewer Notes

💡 This PR clarifies data ordering for user-supplied `knn_graph: csr_matrix` input. It doesn't cover broader updates or additions to the FAQs. A follow-up PR will dive into detailed code examples and discussions on adding or creating your own `knn_graph` inputs. This clarification is an initial step to clear up data ordering requirements before diving deeper into future updates.

### CODE REVIEW

File: cleanlab/datalab/datalab.py [hunk 1]
```diff
@@ -185,7 +185,8 @@ def find_issues(
             evenly distributed across the rows.
             The non-zero entries must be the distances between the corresponding examples. Self-distances must be omitted
             (i.e. the diagonal must be all zeros and the k nearest neighbors of each example must not include itself).
-
+            The data must be sorted in row-major order, and within each row, the values must be sorted in ascending order.
```
Comment (jwmueller) [inline cleanlab/datalab/datalab.py:188]: I think this needs 1-2 sentences more clarification.  What does it mean the data must be sorted? I think readers would interpret "the data" as the original dataset whose ordering is out of our control.

File: cleanlab/datalab/datalab.py [hunk 2]
```diff
@@ -179,12 +179,26 @@ def find_issues(
...
-            Sparse matrix representing distances between examples in the dataset in a k nearest neighbor graph.
+            Sparse matrix of precomputed distances between examples in the dataset in a k nearest neighbor graph.
 
             If provided, this must be a square CSR matrix with shape (num_examples, num_examples) and (k*num_examples) non-zero entries (k is the number of nearest neighbors considered for each example)
-            evenly distributed across the rows.
+            evenly distributed across the rows in a particular format.
             The non-zero entries must be the distances between the corresponding examples. Self-distances must be omitted
             (i.e. the diagonal must be all zeros and the k nearest neighbors of each example must not include itself).
+            Within each row of this matrix, the non-zero entries (distances) must be sorted in ascending order, and the CSR column indices should reflect this ordering of the distances.
+            This type of matrix is returned by the method: `sklearn.`TODO
```
Comment (jwmueller) [inline cleanlab/datalab/datalab.py:189]: note I left TODO here for you

Comment (jwmueller) [reply cleanlab/datalab/datalab.py:189]: also edited the text so `git pull` before resuming work

File: cleanlab/datalab/internal/issue_finder.py [hunk 3]
```diff
@@ -119,6 +119,20 @@ def find_issues(
             evenly distributed across the rows.
             The non-zero entries must be the distances between the corresponding examples. Self-distances must be omitted
             (i.e. the diagonal must be all zeros and the k nearest neighbors of each example must not include itself).
+            Within each row, the non-zero entries (distances) must be sorted in ascending order, and the CSR column indices should reflect this order of distances. This sorting ensures correct neighbor relationships for subsequent analyses.
```
Comment (jwmueller) [inline cleanlab/datalab/internal/issue_finder.py:122]: I think the entire docstring here should just refer reader to the docstring of `knn_graph` argument in `Datalab.find_issues()` with link. No need to repeat the full docstring which is huge

Comment (elisno) [reply cleanlab/datalab/internal/issue_finder.py]: Done.

File: cleanlab/datalab/datalab.py (hunk 2 above)
Comment (elisno) [reply cleanlab/datalab/datalab.py]: Added links both to the [NearestNeighbors.kneighbors_graph method (from sklearn)](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn.neighbors.NearestNeighbors.radius_neighbors_graph), as well as detailed [docs for a `csr_matrix` in scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html).

File: cleanlab/datalab/datalab.py (hunk 1 above)
Comment (elisno) [reply cleanlab/datalab/datalab.py]: This explanation needed a code example and a more explicit data structure description.
I hope the latest addition make things clearer.

File: cleanlab/datalab/internal/issue_finder.py [hunk 4]
```diff
@@ -230,18 +230,10 @@ def find_issues(
...
-            The non-zero entries must be the distances between the corresponding examples. Self-distances must be omitted
-            (i.e. the diagonal must be all zeros and the k nearest neighbors of each example must not include itself).
-
-            For any duplicated examples i,j whose distance is 0, there should be an *explicit* zero stored in the matrix, i.e. ``knn_graph[i,j] = 0``.
-
-            If both `knn_graph` and `features` are provided, the `knn_graph` will take precendence.
-            If `knn_graph` is not provided, it is constructed based on the provided `features`.
-            If neither `knn_graph` nor `features` are provided, certain issue types like (near) duplicates will not be considered.
+            For a detailed description of the formatting requirements, structure, and considerations for the `knn_graph` argument,
+            please refer to the :py:class:`Datalab.find_issues <cleanlab.datalab.datalab.Datalab.find_issues>` documentation.
```
Comment (jwmueller) [inline cleanlab/datalab/internal/issue_finder.py:236]:
```suggestion
            For details, refer to the documentation of the same argument in :py:class:`Datalab.find_issues <cleanlab.datalab.datalab.Datalab.find_issues>`
```

File: cleanlab/datalab/datalab.py [hunk 5]
```diff
@@ -196,19 +196,55 @@ def find_issues(
             If provided, this must be a 2D array with shape (num_examples, num_features).
 
         knn_graph :
-            Sparse matrix representing distances between examples in the dataset in a k nearest neighbor graph.
+            Sparse matrix of precomputed distances between examples in the dataset in a k nearest neighbor graph.
 
-            If provided, this must be a square CSR matrix with shape (num_examples, num_examples) and (k*num_examples) non-zero entries (k is the number of nearest neighbors considered for each example)
+            If provided, this must be a square CSR matrix with shape (num_examples, num_examples) and (k*num_examples) non-zero entries (k is the number of nearest neighbors considered for each example),
```
Comment (jwmueller) [inline cleanlab/datalab/datalab.py:201]:
```suggestion
            If provided, this must be a square CSR matrix with shape ``(num_examples, num_examples)`` and ``(k*num_examples)`` non-zero entries (``k`` is the number of nearest neighbors considered for each example),
```

File: cleanlab/datalab/datalab.py [hunk 6]
```diff
@@ -196,19 +196,55 @@ def find_issues(
...
+            - `indices`: This array stores the column indices of the non-zero elements in the matrix. Each entry in `indices` corresponds to an entry in `data`, indicating the column position of that element.
+            - `indptr`: This array indicates the start and end indices in `data` for each row of the matrix. For example, the non-zero elements of the i-th row are stored from `data[indptr[i]]` to `data[indptr[i+1]]`.
+
+            The matrix represents distances, with each non-zero entry being a distance between examples. Self-distances must be omitted
+            (i.e. the diagonal must be all zeros and the k nearest neighbors for each example do not include the example itself).
+
+            For the knn_graph, within each conceptual "row" of the matrix (as defined by the ranges in `indptr`), the non-zero entries (distances) must be sorted in ascending order. This sorting applies to the segments of the `data` array that correspond to each row. The `indices` array should reflect this ordering, maintaining the correct column positions for these sorted distances.
```
Comment (jwmueller) [inline cleanlab/datalab/datalab.py:212]:
```suggestion
            Each non-zero entry in this matrix is a distance between a pair of examples in the dataset. Self-distances must be omitted
            (i.e. diagonal must be all zeros, k nearest neighbors for each example do not include the example itself).

            This CSR format uses three 1D arrays (`data`, `indices`, `indptr`) to store a 2D matrix ``M``:

            - `data`: 1D array containing all the non-zero elements of matrix ``M``, listed in a row-wise fashion (but sorted within each row).
            - `indices`: 1D array storing the column indices in matrix ``M`` of these non-zero elements. Each entry in `indices` corresponds to an entry in `data`, indicating the column of ``M`` containing this entry.
            - `indptr`: 1D array indicating the start and end indices in `data` for each row of matrix ``M``. The non-zero elements of the i-th row of ``M`` are stored from ``data[indptr[i]]`` to ``data[indptr[i+1]]``.

            Within each row of matrix ``M`` (defined by the ranges in `indptr`), the corresponding non-zero entries (distances) of `knn_graph` must be sorted in ascending order (specifically in the segments of the `data` array that correspond to each row of ``M``). The `indices` array must also reflect this ordering, maintaining the correct column positions for these sorted distances.
```

**Review by jwmueller — APPROVED** [review]
made minor suggestions, but LGTM

File: cleanlab/datalab/internal/issue_finder.py [hunk 7]
```diff
@@ -230,18 +230,9 @@ def find_issues(
             If provided, this must be a 2D array with shape (num_examples, num_features).
 
         knn_graph :
-            Sparse matrix representing distances between examples in the dataset in a k nearest neighbor graph.
+            Sparse matrix representing distances between examples in a dataset in a k nearest neighbor graph.
```
Comment (jwmueller) [inline cleanlab/datalab/internal/issue_finder.py:233]:
```suggestion
            Sparse matrix representing distances between examples in the dataset in a k nearest neighbor graph.
```

**Pull request merged**